drivers/hmis/app/models/hmis/form/custom_data_element_generator.rb

```diff
@@ -0,0 +1,182 @@
+          end
+
+          custom_field_key = item.mapping&.custom_field_key
+
+          # Proceed unless item specifies a custom_field_key mapping, OR create_missing_mappings is true.
+          # @create_missing_mappings will mutate the definition to add new mappings.
+          next unless custom_field_key || @create_missing_mappings
+
+          # Determine the owner type for the CDED
+          owner_type = determine_owner_type(item)
+          # Use referenced key for CDED if present, otherwise generate a new unique key based on link_id
+          cded_key = custom_field_key || ensure_unique_key("#{cded_key_prefix}_#{item.link_id}", owner_type: owner_type)
+
+          @cdeds << Hmis::Hud::CustomDataElementDefinition.new(
+            key: cded_key,
+            label: generate_cded_field_label(item),
+            repeats: item.repeats || false,
+            field_type: infer_cded_field_type(item),
+            owner_type: owner_type,
+            **cded_attributes,
+          )
+
+          # Modify the definition JSON to reference the new CDED key
+          item_hash['mapping'] = { 'custom_field_key' => cded_key }
+        end
+
+        @cdeds
+      end
+
+      private
+
+      # Generate appropriate 'label' for CDED based on item properties.
+      def generate_cded_field_label(item)
+        # Select the first non-blank label from the item properties. Prefers labels that are typically shorter.
+        label = [
+          item.readonly_text,
+          item.brief_text,
+          item.text,
+          item.link_id.humanize,
+        ].compact_blank.first
+
+        ActionView::Base.full_sanitizer.sanitize(label)[0..100].strip
+      end
+
+      # Infer CDED field_type based on Form Item type.
+      def infer_cded_field_type(item)
+        case item.type
+        when 'STRING', 'TEXT', 'CHOICE', 'TIME_OF_DAY', 'OPEN_CHOICE'
+          'string'
+        when 'BOOLEAN'
+          'boolean'
+        when 'DATE'
+          'date'
+        when 'INTEGER'
+          'integer'
+        when 'CURRENCY'
+          'float'
+        when 'FILE', 'IMAGE'
+          'file'
+        else
+          raise "unable to determine cded type for #{item.type}"
+        end
+      end
+
+      def skip_item?(item)
+        Hmis::Form::Definition::NON_QUESTION_ITEM_TYPES.include?(item.type) ||
+          item.mapping&.field_name
+      end
+
+      def lookup_mapped_cded(item)
+        custom_field_key = item.mapping&.custom_field_key
+        return unless custom_field_key
+
+        owner_type = determine_owner_type(item)
+        found = Hmis::Hud::CustomDataElementDefinition.find_by(owner_type: owner_type, key: custom_field_key, data_source: @data_source)
+        raise "Form '#{@definition.identifier}' item '#{item.link_id}' references a CDED '#{custom_field_key}' belongs to a different form ('#{found.form_definition_identifier}')" if found&.form_definition_identifier.present? && found&.form_definition_identifier != @definition.identifier
```

Comment:
Do we still have any forms with cdeds kicking around from before we made the decision that forms should not share cdeds? I remember we did used to have some examples of this. If so, I think this will cause them to always fail on publish.

Reply:
It's OK for forms to share CDEDs. For example we have lots of Service forms that share a "service note" field in one environment, which was intentional for reporting purposes. In those cases, though, the referenced CDED should _not_ have a form_definition_identifier specified. As long as that's true, this_ shouldn't raise, because `found&.form_definition_identifier.present?` will evaluate to false.

I will look into whether it's possible that we have cross-referenced CDEDs  that _do_ have form identifiers listed. I think (hope) those would fail validation.